### PR TITLE
Run crossgen in parallel in crossgen_comparison.py

### LINKED ISSUE #34101: Improve cross-crossgen run time
Our cross-crossgen jobs take ~10 minutes to complete, and hold one linux arm machine for the entire time. The run time can be improved by parallelizing the crossgen invocations. This should crossgen in roughly ~4x the time due to the machines having four cores.

This will hopefully take our job time down to ~5 minutes.

### PR DESCRIPTION
**Pull request by jashook**
This will hopefully reduce the runtime of crossgen_compairision.py and address #33172

### CODE REVIEW

File: src/coreclr/tests/scripts/crossgen_comparison.py [hunk 1]
```diff
@@ -151,6 +152,95 @@ def build_argument_parser():
...
+
+        if self.verbose:
+            print_prefix = "[{}:{}]: ".format(index, size)
+
+        await async_callback(print_prefix, item, *extra_args)
+
+        # Add back to the queue, incase another process wants to run.
+        self.subproc_count_queue.put_nowait(subproc_id)
+
+    async def __run_to_completion__(self, async_callback, *extra_args):
```
Comment (lpereira) [inline src/coreclr/tests/scripts/crossgen_comparison.py]: Is this a dunder method?

File: src/coreclr/tests/scripts/crossgen_comparison.py [hunk 2]
```diff
@@ -544,11 +634,15 @@ def crossgen_framework(args):
...
 
-    for assembly_name in g_Framework_Assemblies:
+    def run_crossgen(print_prefix, assembly_name):
         il_filename = os.path.join(args.core_root, assembly_name)
         ni_filename = os.path.join(ni_files_dirname, add_ni_extension(assembly_name))
         crossgen_results = run_crossgen(args.crossgen_executable_filename, il_filename, ni_filename, platform_assemblies_paths, debugging_files_dirname)
         save_crossgen_results_to_json_files(crossgen_results, args.result_dirname)
+
+    helper = AsyncSubprocessHelper(g_Framework_Assemblies)
+    helper.run_to_completion(run_pmi)
```
Comment (echesakovMSFT) [inline src/coreclr/tests/scripts/crossgen_comparison.py]: Typo: `run_pmi` -> `run_crossgen`

File: src/coreclr/tests/scripts/crossgen_comparison.py (hunk 1 above)
Comment (jashook) [reply src/coreclr/tests/scripts/crossgen_comparison.py]: This is incorrect as mentioned, however, this code already exists in our tree. Leaving as is

**Comment by jashook** [comment]
I believe this is finally ready for review @echesakovMSFT @dotnet/runtime-infrastructure ptal

File: eng/pipelines/common/platform-matrix.yml [hunk 3]
```diff
@@ -38,7 +38,7 @@ jobs:
       archType: arm
       platform: Linux_arm
       container:
-        image: ubuntu-16.04-cross-14.04-23cacb0-20200121150126
+        image: ubuntu-18.04-cross-arm-16.04-09ec757-20200324125113
```
Comment (jaredpar) [inline eng/pipelines/common/platform-matrix.yml]: Why did we need to upgrade the image here?

File: eng/pipelines/coreclr/templates/crossgen-comparison-job.yml [hunk 4]
```diff
@@ -116,13 +116,15 @@ jobs:
       - script: |
           mkdir $(workItemDirectory)\log\$(crossFlavor)
         displayName: Create directories
+
```
Comment (jaredpar) [inline eng/pipelines/coreclr/templates/crossgen-comparison-job.yml]: Nit: extra blank line

**Review by echesakov — APPROVED** [review]
Changes look good to me. I wonder what speedup we will get 
Thank you @jashook !

**Comment by jashook** [comment]
Will post runtime improvements once the run finishes

File: eng/pipelines/coreclr/templates/crossgen-comparison-job.yml [hunk 5]
```diff
@@ -116,13 +116,15 @@ jobs:
...
         displayName: Create directories
+        
 
     # Create baseline output on the host (x64) machine
     - task: PythonScript@0
       displayName: Create cross-platform crossgen baseline
       inputs:
         scriptSource: 'filePath'
         scriptPath: $(coreClrRepoRoot)/tests/scripts/crossgen_comparison.py
+        pythonInterpreter: /usr/bin/python3
```
Comment (jaredpar) [inline eng/pipelines/coreclr/templates/crossgen-comparison-job.yml]: Do we need to add `python3` to our linux requirements markdown file?

File: eng/pipelines/common/platform-matrix.yml (hunk 3 above)
Comment (jashook) [reply eng/pipelines/common/platform-matrix.yml]: The script needs python3.6+

File: eng/pipelines/coreclr/templates/crossgen-comparison-job.yml (hunk 5 above)
Comment (jashook) [reply eng/pipelines/coreclr/templates/crossgen-comparison-job.yml]: We should be one python3, seeing that python2 is eol. We have not exactly mandated it yet.

File: eng/pipelines/common/platform-matrix.yml (hunk 3 above)
Comment (jashook) [reply eng/pipelines/common/platform-matrix.yml]: Its also generally not needed for us to target 14.04 anymore as 14.04 is eol

**Comment by jashook** [comment]
## Time in seconds

| Branch | x64-arm crossgen (build machine) | arm crossgen (on device) |
| ------- | ----------------------------------- | -------------------------- |
| master | 189.140369                                          |  448.044597                       |
| change | 87.963722                                           | 133.222282                          |

**Comment by jashook** [comment]
This will leave our cross-crossgen job with ~3 minutes of helix setup time (download, zip and docker) and ~2 minutes of execution time. Total time is reduced ~40% with this change.

**Comment by trylek** [comment]
Very nice. That matches the ratio between Crossgen / Crossgen2 framework compilation times in "build-test generatelayoutonly" (where there's a straight linear loop over the assemblies) and SuperIlc (which uses parallelization much like you do, just written in C#), something like 3:1, almost reaching the 4:1 ideal speedup on my 4-core box.

**Review by trylek — APPROVED** [review]
Looks great, thanks for pulling off such a dramatic perf improvement!

**Pull request merged**